File: leetcode/0005_最长回文子串/solution.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>

using namespace  std;

// 动态规划
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n));
        string ans;

        for (int l = 0; l < n; l++) {
            for (int i = 0; i + l < n; i++) {
                int j = i + l;
                if (l == 0) {
                    dp[i][j] = 1;
                } else if (l == 1)  {
                    dp[i][j] = (s[i] == s[j]);
                } else {
                    dp[i][j] = (s[i] == s[j] && dp[i + 1][j - 1]);
                }
                if (dp[i][j] && ans.size() < l + 1) {
                    ans = s.substr(i, l + 1);
                }
            }
        }
        return ans;
    }
};

// 中心扩散

class Solution2 {
public:
    pair<int, int> expandAroundCenter(const string& s, int left, int right) {
        while (left >= 0 && right < s.size() && s[left] == s[right]) {
            left--;
            right++;
        }
        return {left + 1, right - 1};
    }
    string longestPalindrome(string s) {
        int start = 0, end = 0;
        for (int i = 0; i < s.size(); i++) {
            auto [left1, right1] = expandAroundCenter(s, i, i);
            auto [left2, right2] = expandAroundCenter(s, i, i + 1);

            if (right1 - left1 > end - start) {
                start = left1;
                end = right1;
            }
            if (right2 - left2 > end - start) {
                start = left2;
                end = right2;
            }
        }
        return s.substr(start, end - start + 1);
    }
};

int main() {
    Solution2 solution;
    string s = "babad";
    string ans =  solution.longestPalindrome(s);
    cout << ans << endl;
    return 0;
}

File: leetcode/0005_最长回文子串/solution.js

const initial2DArray = (n, m, val) => Array(n).fill().map(() => Array(m).fill(val))

var longestPalindrome = function(s) {
    const n = s.length
    const dp = initial2DArray(n, n, false)
    let ans = ''

    for (let l = 0; l < n; l++) {
        for (let i = 0; i + l < n; i++) {
            let j = i + l
            if (l == 0) {
                dp[i][j] = true
            } else if (l == 1) {
                dp[i][j] = (s[i] == s[j])
            } else {
                dp[i][j] = (s[i] == s[j] && dp[i + 1][j - 1])
            }
            if (dp[i][j] && l + 1 > ans.length) {
                ans = s.substr(i, l + 1)
            }
        }
    }

    return ans
};